business/workloads.go

```diff
@@ -170,8 +184,63 @@ func (in *WorkloadService) GetPod(namespace, name string) (*models.Pod, error) {
 	return &pod, nil
 }
 
-func (in *WorkloadService) GetPodLogs(namespace, name string, opts *core_v1.PodLogOptions) (*kubernetes.PodLogs, error) {
-	return in.k8s.GetPodLogs(namespace, name, opts)
+func (in *WorkloadService) getParsedLogs(namespace, name string, opts *core_v1.PodLogOptions) (*PodLog, error) {
+	podLog, err := in.k8s.GetPodLogs(namespace, name, opts)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Matches an ISO8601 full date
+	timestampRegexp := regexp.MustCompile("^\\d{4}-\\d\\d-\\d\\dT\\d\\d:\\d\\d:\\d\\d(\\.\\d+)?(([+-]\\d\\d:\\d\\d)|Z)?(\\+\\d{4})?")
+	severityRegexp := regexp.MustCompile("(\\s+|^)(INFO|WARN|DEBUG|TRACE)\\s+")
+
+	lines := strings.Split(podLog.Logs, "\n")
+	messages := make([]LogMessage, 0)
+
+	for _, line := range lines {
+		message := LogMessage{
+			Message:   line,
+			Timestamp: 0,
+			Severity:  "",
+		}
+
+		timestamp := timestampRegexp.FindString(line)
+		if timestamp != "" {
+			parsed, _ := time.Parse("2006-01-02T15:04:05+0000", string(timestamp))
+
+			if err == nil {
+				message.Timestamp = parsed.Unix()
+			}
+		}
+
+		line = string(timestampRegexp.ReplaceAll([]byte(line), []byte("")))
+		message.Message = line
+
+		severity := severityRegexp.FindString(line)
+		if severity != "" {
+			message.Severity = strings.TrimSpace(severity)
+
+			line = string(severityRegexp.ReplaceAll([]byte(line), []byte("")))
+			line = strings.TrimSpace(line)
+		}
+
+		line = strings.TrimSpace(line)
+		message.Message = line
+
+		messages = append(messages, message)
+	}
+
+	message := PodLog{
+		Logs:     podLog.Logs,
+		Messages: messages,
+	}
+
+	return &message, err
+}
+
+func (in *WorkloadService) GetPodLogs(namespace, name string, opts *core_v1.PodLogOptions) (*PodLog, error) {
```

Comment:
Although our API is not supported I'm not sure we should break the client by changing the return structure for the API call.  I think perhaps we should either have two calls, one that returns the logs as a string, like today, and one that returns the new format.  Or, maybe we shoud have an option (queryParam) that specifies the desired return format, defaulting to today's string.

What do you think?

Reply:
I don't really agree, no, as we do not break the compatibility, the client code can stay the same. Also this means that a partial integration is possible, before commiting to adding the changes to the backend.

Reply:
OK, so I see now what you are doing.  We can temporarily return both the logs blob ('Logs') and log entries ('Entries') until we support only the entries on the client side.  I can't really see the need to maintain the blob format as the client could reconstruct lines or the whole blob as needed.